Add support for Child Blocks

noisysocks opened the pull request:
## Description

Closes #5540. I split this out from https://github.com/WordPress/gutenberg/pull/6346 which was mostly experimental — head there for extra context!

Adds `parent` which lets you restrict a block so that it is only available as a nested block. 

```js
registerBlockType( 'acme/buy-button', {
	title: 'Buy Button',
	icon: 'cart',
	category: 'common',

	// Only allow in products:
	parent: [ 'acme/product' ],

	...
} );
```

## How has this been tested?
I wrote a [test plugin](https://gist.github.com/noisysocks/9b9503bd6489ffa51088d35c7a2a8ba0) which uses the new API. I then made sure that _Buy Button_ blocks only appear when inserted into a _Product_ block.

A quick way to test this is to paste the contents of [`block.js`](https://gist.github.com/noisysocks/9b9503bd6489ffa51088d35c7a2a8ba0#file-block-js) into your browser console.

## Screenshots <!-- if applicable -->

![child-blocks](https://user-images.githubusercontent.com/612155/40043490-ee6a50b6-5867-11e8-824b-1fc87a1c275d.gif)

## Types of changes

Block registration API changes:

- Introduces `parent` which lets one specify which block types a block can be nested into.

Selector changes:

- Adds `canInsertBlockType` which is the source of truth for whether a block can be inserted or nested.
- Removes the `allowedBlockTypes` argument from `getInserterItems`. This can be determined from state, now.
- Items returned by `getInserterItems` will include new `utility` and `frecency` attributes. `utility` indicates how useful we think inserting the block will be. This lets us show contextually useful blocks at the top of the Suggested tab.
- Deprecates `getSupportedBlockTypes`. `getInserterItems` will remove any blocks that are not supported. Alternatively, `canInsertBlockType` can be called directly.
- Deprecates `getFrecentInserterItems`. Instead, call `getInserterItems` and filter/sort by `frecency`.

Inserter changes:

- Both `inserter` and `inserter-with-shortcuts` have been changed to use `getInserterItems` and its new functionality.

noisysocks committed:
Support Child Blocks

Block registration API changes:

- Introduces `parent` which lets one specify which block types a block
  can be nested into.

Selector changes:

- Adds `canInsertBlockType` which is the source of truth for whether a
  block can be inserted or nested.
- Removes the `allowedBlockTypes` argument from `getInserterItems`. This
  can be determined from state, now.
- Items returned by `getInserterItems` will include new `utility` and
  `frecency` attributes. `utility` indicates how useful we think inserting
  the block will be. This lets us show contextually useful blocks at the
  top of the Suggested tab.
- Deprecates `getSupportedBlockTypes`. `getInserterItems` will remove
  any blocks that are not supported. Alternatively, `canInsertBlockType`
  can be called directly.
- Deprecates `getFrecentInserterItems`. Instead, call `getInserterItems`
  and filter/sort by `frecency`.

Inserter changes:

- Both `inserter` and `inserter-with-shortcuts` have been changed to use
  `getInserterItems` and its new functionality.

jorgefilipecosta commented on editor/store/selectors.js:
```diff
@@ -1202,152 +1202,268 @@ export function getNotices( state ) {
+	( state, blockName, parentUID ) => [
+		state.blockListSettings[ parentUID ],
+		state.editor.present.blocksByUID[ parentUID ],
+		state.settings,
```
I think we may be more granular when passing our cache variables and use the selectors to compute the references e.g:[ getEditorSettings( state ).allowedBlockTypes, getBlockListSettings( state, parentUID )... ] this way if the state changes as long as selectors we depend on are updated we are safe :)

noisysocks replied:
Good catch. I've updated the `settings` dependency to be more specific in a60dc232e.

I'm not sure it's a good idea to call a selector from the second argument of `createSelector`. If we invalidated the cache based using `getBlockListSettings( state, parentUID )` and then someone went ahead and changed `getBlockListSettings` to use `createSelector` then it would mean that our cache would never invalidate.

It looks like @aduth implemented a really good solution to this problem in the form of [`getDependants`](https://github.com/aduth/rememo#api) but we are not using version 3.0.0 of rememo yet.

I'll look into updating our version of rememo in a seperate PR so that we can clean this up.

jorgefilipecosta commented on editor/store/selectors.js:
```diff
@@ -1202,152 +1202,268 @@ export function getNotices( state ) {
-	if ( ! allowedBlockTypes || ! blockType ) {
-		return null;
-	}
+export const canInsertBlockType = createSelector(
```
Given that the name of the selector is canInsertBlockTyle I think the selector may also need to take into consideration locking. If a CPT sets a template lock we may not be able to insert a block because of that.

noisysocks replied:
Fair enough! Added in c3a171ad9.

mtias commented:
Great work here. I think we could refine the inserter UI further but we should wait for the results of the design explorations.

Thinking more about the API footprint, I think we should go with `parent` as the name, and split post type support to `postType`.

A few reasons for the first:
- The main point of this addition is to set a relationship between blocks. The term "allowed" is too specific to the insertion mechanism, which is just one aspect of the child-parent relationship.
- "Allowed" makes more sense as an instance specific prop for `InnerBlocks` but less so for a configuration aspect that's part of the block type definition.

For the second, even if I think there's merit in combining, there's enough distinction on post types to go the cautious route — easier to add support later within `parent` than to remove it.

brandonpayton commented on editor/components/inserter-with-shortcuts/index.js:
```diff
@@ -18,18 +18,40 @@ import { withDispatch, withSelect } from '@wordpress/data';
+	 *
+	 * @see selectors.getInserterItems()
+	 */
+	return orderBy( items, [ 'utility', 'frecency' ], [ 'desc', 'desc' ] );
```
There are multiple components sorting by `utility` and `frecency`, and I'd like to provide the same sort in the block completer (#6067) for a consistent insertion experience. Is there a good place we could share this knowledge rather than duplicating?

noisysocks replied:
How do you feel about `getInserterItems` doing the sorting by default? `InserterMenu` would have to disable this though so that we don't double sort.

brandonpayton replied:
I think that would mask our intention for the inserters to use the same default sort and introduce complexity by requiring a way to disable default sort. What do you think of exporting a default sort function from `editor/components/inserter` and having other inserter modules explicitly import it? Then the code would say what we mean which is "Other insertion-related components use the same default sort as the primary insertion menu".

noisysocks replied:
I made `getInserterItems` always sort the items in e9fbe66f5. It turns out that my claim that we'd have to make `InserterMenu` disable this was flat out wrong. Having the selector sort the items makes sense to me because:

- It's basically always what you want to do
- It's more performant to sort the items within `createSelector` where the results are cached

brandonpayton commented on docs/reference/deprecated.md:
```diff
@@ -4,6 +4,9 @@ Gutenberg's deprecation policy is intended to support backwards-compatibility fo
  - `wp.blocks.withEditorSettings` is removed. Please use the data module to access the editor settings `wp.data.select( "core/editor" ).getEditorSettings()`.
+- `getInserterItems`: The `allowedBlockTypes` argument was removed and the `parentUID` argument was added.
+- `getFrecentInserterItems` selector removed. Please use `getInserterItems` instead.
+- `getSupportedBlocks` selector removed. Please use `canInsertBlockType` instead.
```
It's great to see this simplification and to be able to avoid the dance of calling `getSupportedBlocks( parentUID )` to have something to pass to `getInserterItems`. 🙇

noisysocks commented:
> Thinking more about the API footprint, I think we should go with parent as the name, and split post type support to postType.

👍 renamed it to `parent` in b46ec6b75. We'll add block type restricting in a later PR.

tofumatt commented:
FYI I'm out until May 28 so feel free to remove me from review if it's blocking stuff.

brandonpayton commented on editor/store/selectors.js:
```diff
@@ -1363,28 +1487,16 @@ function getItemsFromInserts( state, inserts, allowedBlockTypes, maximum = MAX_R
-				return count / 4;
-		}
-	};
+	deprecated( 'getFrecentInserterItems', {
```
Thanks! If you're interested, we just added support for an optional `hint` message in https://github.com/WordPress/gutenberg/commit/3fd06df7947de492b6708f2532bbaaf34704b948.

noisysocks commented:
How are we feeling about this? Would like to get it merged soon so that it has time to bake before 3.0.

mtias commented:
Let's get this in shortly after doing the minor release fixes.

jorgefilipecosta commented on editor/store/selectors.js:
```diff
@@ -1202,152 +1202,280 @@ export function getNotices( state ) {
+ *                                            the number of inserts that have occurred.
+ */
+function getInsertUsage( state, id ) {
+	return state.preferences.insertUsage[ id ] || { time: undefined, count: 0 };
```
Minor: Should we add a DEFAULT_INSERT_USAGE object equal to { time: undefined, count: 0 };? We are creating a new object each time the id is not found. If later we use this as a prop we will cause rerenders.

noisysocks replied:
I'll just avoid this fanciness altogether and return `null` if the insert is missing 🙂 b2f8cc154

gziolo commented on editor/store/test/selectors.js:
```diff
@@ -2616,224 +2639,335 @@ describe( 'selectors', () => {
+			expect( getInserterItems( state ) ).toEqual( [
 				{
-					id: 'core/block/123',
+					id: 'core/test-block-b',
```
When looking only at this test it is unclear why Block B is listed before Block A. `utility` value inherited from the `common` category does the trick. Maybe Block B should be removed from this test and included in new test which tests only order of the items based on `utility` and `frecency`?

noisysocks replied:
👍 changed in 470d32e4b

gziolo commented on editor/store/test/selectors.js:
```diff
@@ -2616,224 +2639,335 @@ describe( 'selectors', () => {
-			] );
+			const items = getInserterItems( state );
+			const testBlockBItem = items.find( ( item ) => item.id === 'core/test-block-b' );
+			expect( testBlockBItem.utility ).toBe( 1 );
```
Should we use constant `MEDIUM` instead of value `1` to better match with the description of the test?

noisysocks replied:
👍 added some constants in 6efa3afd2

gziolo commented on editor/store/test/selectors.js:
```diff
@@ -2616,224 +2639,335 @@ describe( 'selectors', () => {
+			};
+			const items = getInserterItems( state );
+			const sharedBlock2Item = items.find( ( item ) => item.id === 'core/test-block-b' );
+			expect( sharedBlock2Item.utility ).toBe( 2 );
```
Value `2` is hard to follow in here, the description says `high`.

noisysocks replied:
👍 added some constants in 6efa3afd2

gziolo commented:
This PR needs rebase, also description is out of the date after `allowedParentBlocks` was renamed to `parent` :)

gziolo commented on editor/store/selectors.js:
```diff
@@ -1202,152 +1202,280 @@ export function getNotices( state ) {
-	const staticItems = getBlockTypes().map( ( blockType ) =>
-		buildInserterItemFromBlockType( state, allowedBlockTypes, blockType )
-	);
+		const calculateUtility = ( category, count, isContextual ) => {
```
`utility` is a bit hard to follow with those numbers used, can we use constants instead or some different way to codify them?

noisysocks replied:
👍 added some constants in 6efa3afd2

jorgefilipecosta commented:
I'm not sure if this topic was already discussed before, I was not able to find an explicit discussion regarding this on the other PR, but if it was discussed sorry to bring the topic again and feel free to ignore the comment.

In 'acme/product' we set allowedBlocks: [ 'core/paragraph', 'core/image' ].
In 'acme/buy-button' we set parent: [ 'acme/product' ].

On the inserter of acme/product we have 3 blocks available 'core/paragraph', 'core/image', and 'acme/buy-button' instead of the two the block explicitly set [ 'core/paragraph', 'core/image' ].
Using the parent in the child changed the restriction on the parent.
I think they are two different restrictions and both restrictions should work so they should intersect.
In that case, the  'acme/product' would just show 'core/paragraph', 'core/image' and it would not be possible to insert 'acme/buy-button' in any block because it only allows one parent and that parent does not allow that block.

It looks like allowedBlocks passed in InnerBlocks on the parent is not being respected, and it may cause problems in the future.
Let's imagine we have another block 'acme/buy-button-promotional' and in that block, we only allow 'acme/product'. So both blocks acme/buy-button and acme/buy-button-promotional can only be inserted in acme/product.
But in a given time on 'acme/product' depending on some configuration (may be changeable in the inspector of 'acme/product') we only want to allow the insert of acme/buy-button or acme/buy-button-promotional. That becomes hard to do because the prop allowedBlocks can not restrict the child blocks 'acme/buy-button' and acme/buy-button-promotional.

To get the same behavior ,we have using an "intersection" approach we should:
In 'acme/product' we set allowedBlocks: [ 'core/paragraph', 'core/image', 'acme/buy-button' ].
In 'acme/buy-button' we set parent: [ 'acme/product' ].

mtias commented:
@jorgefilipecosta the current behaviour is ok to me. Allowed blocks is not a strong contract, it's just filtering out blocks from the initial set. Child blocks are always included. If you don't want to include a child, you should unregister the block or filter that block out.

noisysocks commented:
> I'm not sure if this topic was already discussed before, I was not able to find an explicit discussion regarding this on the other PR

For reference, it came up in https://github.com/WordPress/gutenberg/issues/5540#issuecomment-380816384 and https://github.com/WordPress/gutenberg/pull/6346#discussion_r187493406.

noisysocks committed:
Make getInsertUsage return null when no insert exists

noisysocks committed:
Unregister core/block in selector tests

noisysocks committed:
Fix typo

noisysocks committed:
More explicitly handle globallyEnabledBlockTypes param

noisysocks committed:
Deprecate old inserter selectors in 3.2, not 3.1

noisysocks committed:
Add explicit test for items being ordered by utility and frecency

noisysocks committed:
Define constants for the various utility levels

noisysocks commented on editor/store/selectors.js:
```diff
@@ -1202,152 +1202,280 @@ export function getNotices( state ) {
+				return count;
+			}
+
+			const duration = Date.now() - time;
```
I think it's fine since _"Now" in this case is the last time the relevant state changed_ is exactly what we want.

Using the greatest time from all of the blocks is a good idea for a pure alternative, but I don't think it's worth the performance penalty we'd get from iterating through every block.

aduth replied:
>I think it's fine since "Now" in this case is the last time the relevant state changed is exactly what we want.

This would have made for a good inline code comment to the next person who maintains this selector.

noisysocks commented on editor/store/test/selectors.js:
```diff
@@ -1463,6 +1484,8 @@ describe( 'selectors', () => {
 				innerBlocks: [],
 			} );
+
+			unregisterBlockType( 'core/meta-block' );
```
`core/meta-block` is only used in one `it()`.

noisysocks commented on editor/store/test/selectors.js:
```diff
@@ -119,7 +138,9 @@ describe( 'selectors', () => {
 
 	afterAll( () => {
-		unregisterBlockType( 'core/test-block' );
+		unregisterBlockType( 'core/test-block-a' );
```
Oops, fixed in 4f9352e50 👍

noisysocks commented on editor/store/selectors.js:
```diff
@@ -1202,152 +1202,280 @@ export function getNotices( state ) {
+ * @property {boolean}  isDisabled        Whether or not the user should be prevented from inserting
+ *                                        this item.
+ * @property {number}   utility           How useful we think this item is, between 0 and 3.
+ * @property {number}   frecency          Hueristic that combines recency and frecency.
```
🤦‍♂️ 99ca557c1

noisysocks commented on editor/store/selectors.js:
```diff
@@ -1603,24 +1719,16 @@ export function getBlockListSettings( state, uid ) {
-	if ( ! supportedNestedBlocks ) {
-		return false;
-	}
+export function getSupportedBlocks( state, uid ) {
```
`getSupportedBlocks` has to accept `globallyEnabledBlockTypes` to be backwards compatible. I removed it from the signature to avoid an unused variable linter warning. I've made this more explicit by just disabling the linter warning in d42964fdd 👍

noisysocks commented on docs/reference/deprecated.md:
```diff
@@ -7,6 +7,9 @@ Gutenberg's deprecation policy is intended to support backwards-compatibility fo
  - All DOM utils in `wp.utils.*` are removed. Please use `wp.dom.*` instead.
  - `isPrivate: true` has been removed from the Block API. Please use `supports.inserter: false` instead.
  - `wp.utils.isExtraSmall` function removed. Please use `wp.viewport.*` instead.
+ - `getInserterItems`: the `allowedBlockTypes` argument was removed and the `parentUID` argument was added.
```
👍 changed in 4eb77c83a

noisysocks commented on editor/store/selectors.js:
```diff
@@ -1202,152 +1202,280 @@ export function getNotices( state ) {
+			[ 'desc', 'desc' ]
+		);
+	},
+	( state, editorAllowedBlockTypes, parentUID ) => [
```
It's used in `InserterWithShortcuts` as well which is always visible.

noisysocks commented on editor/store/selectors.js:
```diff
@@ -1603,24 +1719,16 @@ export function getBlockListSettings( state, uid ) {
-	}
+export function getSupportedBlocks( state, uid ) {
+	deprecated( 'getSupportedBlocks', {
+		version: '3.1',
```
👍 changed in 4eb77c83a

noisysocks commented on editor/store/selectors.js:
```diff
@@ -1202,152 +1202,280 @@ export function getNotices( state ) {
-function fillWithCommonBlocks( inserts ) {
-	// Filter out any inserts that are associated with a block type that isn't registered
-	const items = inserts.filter( ( insert ) => getBlockType( insert.name ) );
+			if ( ! canInsertBlockType( state, blockType.name, parentUID ) ) {
```
Where are you seeing these? When I run `npm run lint` there's no warnings 😕 

(Also, I like the explicitness…)

aduth replied:
I disagree here, as it seems to imply we expect the result of `canInsertBlockType` to be anything other than a boolean.

- If we do expect it to be a boolean, why not just return it as-is?
- If we don't expect it to be a boolean, is a "can" prefix a good choice of naming?

noisysocks commented:
Thanks for reviewing @jorgefilipecosta and @gziolo! I've rebased (I squashed first to make my life easier), updated the PR description, and replied to your comments inline.

jorgefilipecosta commented:
Thank you for the clarification @noisysocks and @mtias!

jorgefilipecosta commented on editor/components/inserter/index.js:
```diff
@@ -93,18 +88,17 @@ export default compose( [
+			getInserterItems,
 		} = select( 'core/editor' );
-		const { allowedBlockTypes, templateLock } = getEditorSettings();
+		const { templateLock } = getEditorSettings();
```
Minor: We can now remove the usage of template lock as getInserterItems if there is a lock getInserterItems will return no items.

noisysocks replied:
Fixed in 4f6e662b1 👌

jorgefilipecosta commented on editor/components/inserter/menu.js:
```diff
@@ -211,22 +223,6 @@ export class InserterMenu extends Component {
-		};
-	} ),
 	withDispatch( ( dispatch ) => ( {
 		fetchSharedBlocks: dispatch( 'core/editor' ).fetchSharedBlocks,
```
Minor: It is a little strange that we are dispatching a fetch action here but we are not selecting anything in this component. To make this relationship more clear maybe the `items: getInserterItems( rootUid )` should continue to be done here and not passed as prop.

noisysocks replied:
I'll leave this for now because we're looking into replacing this action with a resolver which would address this.

aduth replied:
>I'll leave this for now because we're looking into replacing this action with a resolver which would address this.

Where is this tracked?

jorgefilipecosta commented on editor/store/selectors.js:
```diff
@@ -1196,152 +1200,280 @@ export function getNotices( state ) {
+					return count * 4;
+				case duration < ( 24 * 3600 ):
+					return count * 2;
+				case duration < ( 7 * 24 * 3600 ):
```
minor: Maybe we can create constants at the top e.g: SECONDS_IN_A_WEEK = 7 * SECONDS_IN_A_DAY.

noisysocks replied:
Fixed in 7e215a048 👌

jorgefilipecosta reviewed (approved):
Manually testing this PR, it looks good to me, I left some non-blocker code improvements.
We will continue to iterate on this, e.g., apply restrictions on autocomplete and transforms, add locking to nesting, etc. So during the next steps, if we find a possible improvement, we can do it in the follow-up PRs.
I think we can merge now, so we have more time to test before the release.
Good work @noisysocks!

noisysocks committed:
Remove unnecessary templateLock prop

getInserterItems() will check whether or not the template is locked, so
we don't need to do it twice.

noisysocks committed:
Use SECONDS_PER_* constants

Make calculateFrecency() more readable by using SECONDS_PER_* constants
instead of magic numbers.

noisysocks commented:
Thanks all! 🎉❤️

aduth commented on editor/store/selectors.js:
```diff
@@ -1196,152 +1203,280 @@ export function getNotices( state ) {
+ * Each item object contains what's necessary to display a button in the
+ * inserter and handle its selection.
+ *
+ * The 'utility' property indicates how useful we think an item will be to the
```
Can you explain the difference between `utility` as defined here and `priority` as defined by [`@wordpress/hooks`](https://npmjs.com/package/@wordpress/hooks) and the [block type `transforms` property](https://wordpress.org/gutenberg/handbook/block-api/#transforms-optional)?

aduth replied:
>Not sure that I understand your question, sorry. Would you prefer that I rename `utility` → `priority`? I named it `utility` because it's a measure of _how **useful** an inserter item is_, but I'm not tied to the name.

I'm curious more in hopes of minimizing the number of concepts which exist, where an existing pattern already exists to serve a need. As I read further, I see that these are used as specific multipliers, so while similar to priority, not quite the same.

Which is to say: Nothing should need to be changed here.

gziolo replied:
At some point we might want to consolidate those two properties: utility and frecency into one property called score. This is how some search engines work. The score could be calculated using [weighted sum model](https://en.m.wikipedia.org/wiki/Weighted_sum_model) to translate how things are sorted at the moment.

aduth commented on editor/store/selectors.js:
```diff
@@ -1202,152 +1202,280 @@ export function getNotices( state ) {
-	if ( ! referencedBlock ) {
-		return null;
-	}
+		if ( ! isBlockAllowedInParent ) {
```
>Eslint says you can do:
>```js
>return isBlockAllowedInParent;
>```

This was never addressed.

I'd go further and say the `isBlockAllowedInParent` variable is unnecessary. We could just `return` at the point where it is assigned above. I don't even think an argument of it being an [explaining variable](https://refactoring.com/catalog/extractVariable.html) is valid, since it's meaning is implied as being the return value of the function (i.e. encapsulated already in the function naming).

aduth commented on editor/store/selectors.js:
```diff
@@ -1196,152 +1203,280 @@ export function getNotices( state ) {
+	( state, blockName, parentUID ) => [
+		state.blockListSettings[ parentUID ],
+		state.editor.present.blocksByUID[ parentUID ],
+		state.settings.allowedBlockTypes,
```
Aside: This pierces the abstraction of `getEditorSettings` in use by the selector. Not an easy problem to solve, though.

aduth commented on editor/store/selectors.js:
```diff
@@ -1196,152 +1203,280 @@ export function getNotices( state ) {
-		return null;
-	}
+export const canInsertBlockType = createSelector(
+	( state, blockName, parentUID = null ) => {
```
Why do we assign a default value for `parentUID` ?

aduth replied:
>So that one can call `wp.data.select( 'core/editor' ).canInsertBlockType( 'block-id' )` to determine if a block is insertable into the post.

My concern is less that it's optional, more about whether `null` is treated specially here. In other words, would anything change if `= null` were removed?

noisysocks replied:
Ah, got it. Nothing would change—`undefined` would work just as well. I tend use ` = null` as a way of explicitly denoting which parameters are optional. This may be because I'm not used to working in codebases with consistent JSDOc comments 🙂

aduth commented on editor/store/selectors.js:
```diff
@@ -1196,152 +1203,280 @@ export function getNotices( state ) {
-	}
+export const canInsertBlockType = createSelector(
+	( state, blockName, parentUID = null ) => {
+		const checkAllowList = ( list, item, defaultResult = null ) => {
```
Did you consider making this part of the `wp.blocks` API instead? Something like `wp.blocks.isValidParentForBlockType`? Otherwise it implies that we expect anyone working with `parent` value to understand and accommodate all of its supported shapes, as we've done here.

noisysocks replied:
I don't think `isValidParentForBlockType` is quite right since this function is also used for checking the `allowed_block_types` global variable and the `allowedBlocks` property of an `InnerBlocks`.

I did consider making this function a `util` but decided not to as:

1. Slight lean towards YAGNI
2. I want to encourage folks to use `canInsertBlockType` instead of rolling their own `parent` checking. `canInsertBlockType` correctly takes into account all of the edge cases e.g. template locks, `allowedBlocks` restrictions, `parent` restrictions , `allowed_block_type` restrictions. We already have a few bugs caused by these edge cases not being handled

aduth commented on editor/store/selectors.js:
```diff
@@ -1196,152 +1203,280 @@ export function getNotices( state ) {
-	}
+export const canInsertBlockType = createSelector(
+	( state, blockName, parentUID = null ) => {
+		const checkAllowList = ( list, item, defaultResult = null ) => {
```
Why is the default result `null`, when it's assigned and used as if it were a boolean, i.e. `hasBlockAllowedParent`?

noisysocks replied:
`null` indicates that there isn't an allow list i.e. the block has no preference on what its parent is or what its children are.

aduth commented on editor/store/selectors.js:
```diff
@@ -1196,152 +1203,280 @@ export function getNotices( state ) {
+ * 3. Blocks that are in the common category (utility = 1)
+ * 4. All other blocks (utility = 0)
+ *
+ * The 'frecency' property is a herustic (https://en.wikipedia.org/wiki/Frecency)
```
Nit: Typo: "herustic" -> "heuristic" 😄

noisysocks replied:
👌 oops! 594bb0971

aduth commented on editor/store/selectors.js:
```diff
@@ -1196,152 +1203,280 @@ export function getNotices( state ) {
-		return [];
-	}
+ *
+ * @typedef {Object} Editor.InserterItem
```
Is `typedef` as part of a function JSDoc valid?

noisysocks replied:
Hard to say. It seems to depend on the implementation. My editor (VS Code) is fine with it. I ran `npx jsdox editor/store/selectors.js` to see what would happen but `jsdox` seems to choke on all of our selectors that use `createSelector`.  I think it's fine to revisit this if/when we look at generating API docs.

aduth commented on editor/store/selectors.js:
```diff
@@ -1196,152 +1203,280 @@ export function getNotices( state ) {
+			}
+
+			const duration = Date.now() - time;
+			switch ( true ) {
```
What's the difference between `switch` as used here and an `if`, `else if`, `else` statement?

noisysocks replied:
They're functionally the same. I went with `switch` because I think it reads a little nicer when the conditions line up.

aduth commented on editor/store/selectors.js:
```diff
@@ -1196,152 +1203,280 @@ export function getNotices( state ) {
+
+			let isDisabled = false;
+			if ( blockType.useOnce ) {
+				isDisabled = getBlocks( state ).some( ( block ) => block.name === blockType.name );
```
FYI: Lodash's `_.some` has a nice convenience for this pattern in its object predicates:

```js
isDisabled = some( getBlocks( state ), { name: blockType.name } );
```

https://lodash.com/docs/4.17.10#some

noisysocks replied:
👌 changed in ed0b4f104

aduth commented on editor/store/selectors.js:
```diff
@@ -1196,152 +1203,280 @@ export function getNotices( state ) {
+
+			const duration = Date.now() - time;
+			switch ( true ) {
+				case duration < SECONDS_PER_HOUR:
```
`Date.now` returns a value measured in milliseconds. Is it correct that these constants are measured in seconds?

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/now

noisysocks replied:
Great catch! Fixed in c3dd54b0e

aduth commented on editor/store/selectors.js:
```diff
@@ -1597,24 +1720,18 @@ export function getBlockListSettings( state, uid ) {
  *
  * @return {string[]|boolean} Blocks that can be nested inside the block with the specified uid, or true/false to enable/disable all types.
  */
+// Disable reason: We have to accept `globallyEnabledBlockTypes` so as to be backwards compatible
```
I don't understand. Can you explain what will break if `globallyEnabledBlockTypes` were removed as an argument here?

noisysocks replied:
Nothing will break, but it masks the fact that this method accepts `globallyEnabledBlockTypes` as its third argument and must continue to do until it is removed in 3.2.

For example, say that we remove the argument. Now, imagine that a developer wants to add a new `sortBy` argument in version 3.0. They would likely change the signature to look like this:

```jsx
function getSupportedBlocks( state, uid, sortBy = 'asc' )
```

The developer assumes that this is a non-breaking change, since they are just adding a new optional argument. But this is incorrect, since it will break any pre-3.0 code that makes calls such as `wp.data.select( 'core/editor' ).getSupportedBlocks( 'abc123', [ 'core/paragraph' ] )`.

aduth replied:
Sounds reasonable. My question was mostly for...

- Ensuring we don't keep dead code around (the reason the rule exists in the first place)
- Clarifying wording, since it implies that it is necessary for some reason, though it is not used ("We have to accept...")

tomasruud commented:
Are there any plans on making it possible to fetch nested block contents from a dynamic block?

aduth commented:
@tomasruud The implementation of `do_blocks` (front-end block processing) is currently incapable of supporting nesting within dynamic blocks, since it does not consider tag balancing for dynamic block markers.

See also: https://github.com/WordPress/gutenberg/pull/4591#issuecomment-362300110

aduth commented on editor/store/selectors.js:
```diff
@@ -1196,152 +1203,280 @@ export function getNotices( state ) {
-		return null;
-	}
+export const canInsertBlockType = createSelector(
+	( state, blockName, parentUID = null ) => {
```
Was there a reason we named this with a "parent" prefix, unlike "root" we use everywhere else?

noisysocks replied:
Yes, I suppose it should be renamed for consistency.

noisysocks replied:
https://github.com/WordPress/gutenberg/pull/11274
